src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -289,32 +289,67 @@ public bool TryModifyEntries(ImmutableArray<T> outputs, IEqualityComparer<T> com
                     return true;
                 }
 
-                var modified = new TableEntry.Builder();
+                // We may be able to move the previous entry over wholesale.  So avoid creating an builder and doing any
+                // expensive work there until necessary.  We can only do this if the counts of before/after are the same.
```

Comment:
>necessary [](http://example.com/codeflow?start=46&length=9)

nit: consider "necessary (we detect a non-cached item or a modified state)"

Reply:
will do.

Reply:
done.